### PR TITLE
add fastText custom model for classification

### PR DESCRIPTION
**Pull request by songqiang**
Added wrapper for fastText model as another baseline model to work with NLP features

*Issue #, if available:* 
N/A

*Description of changes:*
Added the fastText custom model in autogluon/utils/tabular/ml/models/fasttext/

Changed Feature Generators to pass through the raw NLP features for fastText model to use. With this change, we can also enable LR model to use text features. May cause memory issue as all text columns will be retained in memory. 

See example notebook in: docs/tutorials/tabular-fasttext.md

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: autogluon/utils/tabular/features/abstract_feature_generator.py [hunk 1]
```diff
@@ -102,7 +102,8 @@ def fit_transform(self, X: DataFrame, y=None, banned_features=None, fix_categori
             if unique_value_count == 1:
                 self.features_to_remove_post.append(column)
             # TODO: Consider making 0.99 a parameter to FeatureGenerator
-            elif 'object' in self.feature_type_family and column in self.feature_type_family['object'] and (unique_value_count / X_len > 0.99):
+            elif ('object' in self.feature_type_family and column in self.feature_type_family['object'] and (unique_value_count / X_len > 0.99) 
+                and column not in self.feature_type_family['text']):
```
Comment (Innixma) [inline autogluon/utils/tabular/features/abstract_feature_generator.py:106]: add `('text' not in self.feature_type_family or column not in self.feature_type_family['text'])` , to avoid case where 'text' doesn't exist in feature_type_family being accessed.

File: autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        text_col = (
+            X[self.text_cols]
+            .astype(str)
+            .fillna(" ")
+            .apply(lambda r: " __NEWCOL__ ".join(v for v in r.values), axis=1)
+            .str.lower()
+            .str.replace("<.*?>", " ")  # remove html tags
+            # .str.replace('''(\\d[\\d,]*)(\\.\\d+)?''', ' __NUMBER__ ') # process numbers preserve dot
+            .str.replace("""([\\W])""", " \\1 ")  # remove special characters
+            .str.replace("\n", " ")
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py:93]: What do you think about replacing "\n" with "\s"?

File: autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# AUTOGENERATED! DO NOT EDIT! File to edit: m.ml.models.fasttext.hyperparameters.parameters.ipynb (unless otherwise specified).
+
+__all__ = ["get_param_baseline"]
+
+# Cell
+# https://fasttext.cc/docs/en/python-module.html#train_supervised-parameters
+def get_param_baseline():
+    params = {
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py:8]: For max-flexibility, we probably want to include all the parameters here, including those like: 
`dim`, `ws`, etc.

Also those like `thread` that don't affect accuracy but affect speed.

File: docs/tutorials/tabular-fasttext.md [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+)
+
+test_data['text'] = (
+    test_data[['education', 'marital-status', 'occupation', 'relationship']]
+    .apply(lambda r: ' '.join(r.values), axis=1)
+)
+
+predictor = task.fit(train_data=train_data, 
+                     label=label_column, 
+                     hyperparameters=custom_hyperparameters)  # Train a single default NaiveBayesModel
```
Comment (jwmueller) [inline docs/tutorials/tabular-fasttext.md:139]: Remove NaiveBayes comment

File: docs/tutorials/tabular-fasttext.md [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+# Use fastText model in TabularPrediction task 
+
+
+```python
+custom_hyperparameters = {'RF': {},
+                         FastTextModel:  {'epoch': 50}
+                         }
+# mock some NLP fields
+train_data['text'] = (
+    test_data[['education', 'marital-status', 'occupation', 'relationship']]
```
Comment (jwmueller) [inline docs/tutorials/tabular-fasttext.md:128]: why is `test_data` used here?

File: docs/tutorials/tabular-fasttext.md [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    To change this, specify the eval_metric argument of fit()
+    AutoGluon will early stop models using evaluation metric: accuracy
+    Custom Model Type Detected: <class 'autogluon.utils.tabular.ml.models.fasttext.fasttext_model.FastTextModel'>
+    Fitting model: RandomForestClassifier ...
+    	0.8588	 = Validation accuracy score
+    	1.19s	 = Training runtime
+    	0.15s	 = Validation runtime
+    Fitting model: FastTextModelClassifier ...
+    NLP features ['text']
+    	0.7556	 = Validation accuracy score
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply docs/tutorials/tabular-fasttext.md:199]: Agreed. Any idea why the accuracy is so poor here? I'd expect 'education', 'occupation' features to provide decent predictive signal.  Maybe you can just choose a different subset of features to join to create a more useful `text` feature for predicting income.

File: docs/tutorials/tabular-fasttext.md [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+                         FastTextModel:  {'epoch': 50}
+                         }
+# mock some NLP fields
+train_data['text'] = (
+    test_data[['education', 'marital-status', 'occupation', 'relationship']]
+    .apply(lambda r: ' '.join(r.values), axis=1)
+)
+
+test_data['text'] = (
+    test_data[['education', 'marital-status', 'occupation', 'relationship']]
```
Comment (jwmueller) [inline docs/tutorials/tabular-fasttext.md:133]: Consider trying more predictive feature-set, like say:
['education', 'occupation', 'workclass', 'native-country']

Comment (jwmueller) [reply docs/tutorials/tabular-fasttext.md:133]: I'd also add some dummy punctation at the end or when concatenating two of the columns, just to show the fastText preprocessing can properly handle this.

Also add:
`print(train_data['text'].head())`
to show user what this `text` field looks like.

File: autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    def preprocess(self, X):
+        text_col = (
+            X[self.text_cols]
+            .astype(str)
+            .fillna(" ")
+            .apply(lambda r: " __NEWCOL__ ".join(v for v in r.values), axis=1)
+            .str.lower()
+            .str.replace("<.*?>", " ")  # remove html tags
+            # .str.replace('''(\\d[\\d,]*)(\\.\\d+)?''', ' __NUMBER__ ') # process numbers preserve dot
+            .str.replace("""([\\W])""", " \\1 ")  # remove special characters
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py:92]: Does this properly handle apostrophes?

File: autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+        "lr": 0.1,
+        "epoch": 10,
```
Comment (jwmueller) [inline autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py:10]: Is 10 too small for default? Will this model tend to overfit?

**Comment by jwmueller** [comment]
Thanks for your PR, fastText will be great to have!

For future PR:  Have you thought about how users might be able specify the `pretrainedVectors` argument to `train_supervised`?
It seems like this could greatly boost performance in many problems, but also requires the user to have the pretrained-vectors file available which may be cumbersome?  More generally using the pretrained fastText to embed text features seems like another great text-preprocessing alternative to have in AutoGluon in addition to its n-gram/BERT featurization, but also requires the user to have the pretrained-vector file (and possibly the file for the relevant language as well)...

File: autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+# Cell
+import pandas as pd
+
+# Cell
+from ....utils.loaders import load_pkl
+from ...constants import BINARY, REGRESSION
+from ..abstract.abstract_model import AbstractModel
+from .hyperparameters.parameters import get_param_baseline
+
+logging.basicConfig(
```
[the remark this answers is not in the document]
Comment (songqiang) [reply autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py:22]: It should not matter. The basicConfig can appear in multiple files, however the settings in the 1st imported file take precedence, which overwrites the settings in any subsequent imported modules. The user can also set it in their main script, which will overwrite any such settings in autogluon.

On the other hand, I do feel the log will be more useful with some additional information. In particular the %(asctime)s prints a timestamp, helping users to quickly see which step take how long to run.

File: autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+    format="%(asctime)s: [%(funcName)s] %(message)s", level=logging.INFO
+)
+logger = logging.getLogger(__name__)
+
+
+def try_import_fasttext():
+    try:
+        import fasttext
+
+        _ = fasttext.__file__
```
[the remark this answers is not in the document]
Comment (songqiang) [reply autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py:32]: I am running some automatic code formatting tools with autoflake, isort and black code formatter. Without this line "_ = fasttext.__file__", fasttext is imported, but not used in the scope of this function, and autoflake will remove the import statement all together.

File: autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+            self._set_default_param_value(param, val)
+
+    def _fit(self, X_train, y_train, **kwargs):
+        import fasttext
+
+        if self.feature_types_metadata is None:
+            from pandas.api.types import is_numeric_dtype
+
+            self.text_cols = [
+                c
```
[the remark this answers is not in the document]
Comment (songqiang) [reply autogluon/utils/tabular/ml/models/fasttext/fasttext_model.py:53]: this is again due to the code auto formatting. Sometimes it is not that intelligent

File: docs/tutorials/tabular-fasttext.md (hunk 5 above)
Comment (songqiang) [reply docs/tutorials/tabular-fasttext.md:128]: Typo. This also explains why the accuracy is only 0.76 below.

File: docs/tutorials/tabular-fasttext.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (songqiang) [reply docs/tutorials/tabular-fasttext.md:199]: The 0.76 accuracy is caused by a typo above. After fixing the typo, RF gets 0.8348, fasttext gets 0.8244, and the ensemble gets 0.8448. 

Do we have an standard datasets with real text columns for demo here?

File: autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py (hunk 9 above)
Comment (songqiang) [reply autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py:10]: I do not have extensive testing of the best default value for number of epochs. The 10 is mostly for not getting too slow on big datasets.

In the income class dataset, here is are some test runs. Maybe I change the default to 50. 

- epoches: 10, train accurary 0.830906, test accurary 0.838878
- epoches: 50, train accurary 0.832339, test accurary 0.841028
- epoches: 100, train accurary 0.832391, test accurary 0.840925
- epoches: 200, train accurary 0.832851, test accurary 0.841335

**Comment by songqiang** [comment]
I resolved most of the comments. Any suggestion on a standard test data that have real NLP columns that I can include in the tutorial?

File: autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py (hunk 9 above)
Comment (jwmueller) [reply autogluon/utils/tabular/ml/models/fasttext/hyperparameters/parameters.py:10]: Thanks for the experiment!  I agree 50 seems like a more reasonable default than 10 although it would be nice to explore values like 20/30 as well with more datasets.

**Comment by jwmueller** [comment]
> I resolved most of the comments. Any suggestion on a standard test data that have real NLP columns that I can include in the tutorial?

Note that your tutorial notebook should be formatted like this:
https://raw.githubusercontent.com/awslabs/autogluon/master/docs/tutorials/tabular_prediction/tabular-quickstart.md
(with none of the code's output included in the .md file)

Re dataset:  you could use some of the standard datasets used in NLP research like:
https://www.kaggle.com/columbine/imdb-dataset-sentiment-analysis-in-csv-format

@sxjscience maybe can share some of the text datasets from GLUE which he has reformatted into CSV already.

File: autogluon/utils/tabular/features/abstract_feature_generator.py [hunk 13]
```diff
@@ -144,6 +145,9 @@ def fit_transform(self, X: DataFrame, y=None, banned_features=None, fix_categori
         self.feature_types_metadata = FeatureTypesMetadata(feature_types_raw=feature_types_raw, feature_types_special=self.feature_type_family_generated)
 
         self.features = list(X_features.columns)
+        for raw_nlp_feature in set(self.feature_type_family['text']) - set(self.features_categorical_final):
+            self.features.remove(raw_nlp_feature)
```
[the remark this answers is not in the document]
Comment (songqiang) [reply autogluon/utils/tabular/features/abstract_feature_generator.py]: I was afraid some downstream code may refer to self.features, which will include the raw text fields. But it seems fine to remove that.

**Comment by szha** [comment]
Job PR-574-2 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-574/2/index.html

**Comment by sxjscience** [comment]
@songqiang @jwmueller

SST in GLUE:
- train: https://autogluon-text.s3-us-west-2.amazonaws.com/glue/sst/train.parquet
- dev: https://autogluon-text.s3-us-west-2.amazonaws.com/glue/sst/dev.parquet
- test: https://autogluon-text.s3-us-west-2.amazonaws.com/glue/sst/test.parquet

STS in GLUE:
- train: https://autogluon-text.s3-us-west-2.amazonaws.com/glue/sts/train.parquet
- dev: https://autogluon-text.s3-us-west-2.amazonaws.com/glue/sts/dev.parquet
- test: https://autogluon-text.s3-us-west-2.amazonaws.com/glue/sts/test.parquet

**Comment by jwmueller** [comment]
Thanks @sxjscience  
The SST dataset is probably more appropriate for evaluating supervised-fasttext, since its model-capacity may be too small to perform reasonably in STS.

**Comment by Innixma** [comment]
I will plan to review once #584 is merged, as it completely overhauls the feature generator.

**Comment by szha** [comment]
Job PR-574-3 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-574/3/index.html

**Comment by szha** [comment]
Job PR-574-4 is done. 
Docs are uploaded to http://autogluon-staging.s3-website-us-west-2.amazonaws.com/PR-574/4/index.html

**Comment by Innixma** [comment]
@songqiang Now that #584 is merged, could you refactor this PR?

Specifically, you should not need to make any changes outside of the model code itself to the repo, and for testing, you can use:

```
from autogluon.utils.tabular.features.generators import PipelineFeatureGenerator, IdentityFeatureGenerator
from autogluon import TabularPrediction as task

feature_generator = PipelineFeatureGenerator(generators=[[IdentityFeatureGenerator(infer_features_in_args=dict(valid_special_types=['text']))]])

task.fit(..., feature_generator=feature_generator, hyperparameters={your_model_hyperparameters})
```

In order to train properly with other models, I need to make a new PR which enables multi-feature generator input to task.fit, but this code should allow you to atleast train the fastText model in isolation.
